File: Algorithms/Searching_Algorithms/Binary_Search/C/Binary_Search_Perfect_Square.c
// Program in C to implement binary search algorithm while finding out whether a
// number is a perfect square or not
#include <stdio.h>

// Recursive function to implement binary search.It returns the number,of which
// the given number is a perfect square. Otherwise,it returns the floor of it's
// square root.(rounded off to the next integer) e.g.,for 100,it will
// return 10.While for 99,it will also return 10.

int perfectSquare(int l, int r, int number) {
  int mid = (l + r) / 2;
  // Binary Search
  while (l <= r) {
    if (l > r) {
      return -1;
    }
    if (mid * mid == number)
      return mid;
    // moves to left half
    else if (mid * mid > number)
      return perfectSquare(l, mid - 1, number);
    // moves to right half
    else
      return perfectSquare(mid + 1, r, number);
  }
}

// Driver program to test the above function
int main() {
  int number = 1225;
  printf("%d", perfectSquare(1, number, number));

  return 0;
}

File: Algorithms/Searching_Algorithms/Binary_Search/Python/Binary_Search_Perfect_Square.py
# Find the square of a given number in a sorted array. If present, Return Yes with the index of element's index
# Else return No

# Program using Recursive Binary Search

# Returns True for Yes and False for No


def BinarySearch(arr, left, right, NumSquare):

    # check for preventing infinite recursive loop
    if right >= left:

        mid = left + (right - left) // 2

        # If element is present at the middle itself
        if arr[mid] == NumSquare:
            return True

        # If element is smaller than mid, then it
        # can only be present in left subarray
        elif arr[mid] > NumSquare:
            return BinarySearch(arr, left, mid - 1, NumSquare)

        # Else the element can only be present
        # in right subarray
        else:
            return BinarySearch(arr, mid + 1, right, NumSquare)

    # Element not present in array
    else:
        return False


# input a sorted array. A sorted array is either of ascending order or descending order
arr = [int(x) for x in input().split()]

left = 0  # left index
Num = int(input())  # Number whose square is to be searched for
NumSquare = Num * Num

# Function CAll
result = BinarySearch(arr, 0, len(arr) - 1, NumSquare)

# check for True
if result:
    print("Yes! " + "The index of square of {} is {}".format(Num, arr.index(Num * Num)))

# For false
else:
    print("No")
